python/ray/serve/tests/unit/test_singleton_thread_router.py

```diff
@@ -0,0 +1,266 @@
+import sys
+import asyncio
+import concurrent.futures
+import time
+from unittest.mock import Mock, patch
+
+import pytest
+
+from ray._private.test_utils import wait_for_condition
+from ray.serve._private.router import SingletonThreadRouter
+from ray.serve._private.common import DeploymentID, RequestMetadata
+from ray.serve._private.replica_result import ReplicaResult
+from ray.serve._private.router import AsyncioRouter
+
+
+def test_request_assignment():
+    mock_replica_result = Mock(spec=ReplicaResult)
+
+    mock_router = Mock(spec=AsyncioRouter)
+
+    async def mock_assign_request(*args, **kwargs):
+        return mock_replica_result
+
+    mock_router.assign_request.side_effect = mock_assign_request
+
+    with patch("ray.serve._private.router.AsyncioRouter", return_value=mock_router):
```

Comment:
I'd suggest dependency injection instead of the mocking. Personally prefer to avoid mocking at all costs, otherwise you are now testing internal implementation details such as the import path of `AsyncioRouter`, which makes updating tests and general refactoring very painful.